### ISSUE TITLE
yielding for completed-signal does not work with more than one yield inside a coroutine

### ISSUE THREAD

**Issue by Warlaan** [godotengine/godot#17280]
This probably isn't as much a bug as maybe a design flaw.

**Godot version:**
3.0.2

**OS/device including version:**
Windows 10 / in editor

**Issue description:**
When a function contains the yield keyword it returns a GDFunctionState. That GDFunctionState in turn has a "completed" signal. So you can run coroutines synchronously using 
`yield(coroutine(), "completed")`

Once a coroutine contains more than one yield-command that doesn't work anymore - I am guessing because the outer yield waits for the first GDFunctionState to emit "completed", but that GDFunctionState is discarded and replaced with the one from the second inner yield.

**Steps to reproduce:**
In this code "ready is over" is printed once the Button is pressed:
```
extends Node

func _ready():
	yield(test(), "completed")
	print("ready is over")

func test():
	yield($Button, "pressed")
```

But here it's never printed because the GDFunctionState returned first by test() does never complete:
```
extends Node

func _ready():
	yield(test(), "completed")
	print("ready is over")

func test():
	yield($Button, "pressed")
	yield($Button2, "pressed")
```

This third version works, but of course it's not a good solution because you would have to know how many times to resume before yielding for the complete-signal:
```
extends Node

func _ready():
	var state = test()
	var state2 = state.resume()
	yield(state2, "completed")
	print("ready is over")

func test():
	yield($Button, "pressed")
	yield($Button2, "pressed")
```

So in case it's not clear: what I am suggesting is that there should be a way to wait for a coroutine to complete. Be it either by changing the behaviour of the yield keyword so the returned GDFunctionState is updated rather than replaced or by introducing a new mechanic similar to C#'s await keyword.

**Comment by Zylann** [comment]
I agree with this. I didn't end up yet in a situation like this in Godot, but I had it many times in other projects, and the few times I experimented with Godot's `yield` I thought this limitation would occur.
Could GDScript support yield enumerators as in C#? This is where I learned the first time how to use `yield`, which I use extensively today, but when I started using `yield` in GDScript, sure it does some job, but I feel it's kind of limited in comparison.

**Comment by neikeq** [comment]
This looks like a regression to me. I'm sure I tested those use cases when I added the "completed" signal to GDFunctionState, since that's the exact purpose I added this signal for. There is even a code snippet in the PR: #8573

**Comment by Warlaan** [comment]
Hopefully it is - but in the PR you linked there's only one code snippet that looks like my first example (with only one yield in the coroutine), and that's the case that works. It's when there are two or more yields in the coroutine that the "completed"-signal doesn't work anymore. Or did you refer to a different code snippet that I missed?

**Comment by Warlaan** [comment]
@Zylann: please keep in mind that GDscript's yield is targeted at coroutines, so it's more like C#'s async and await keywords than its yield keyword. The yield keyword is only used as a workaround in Unity's C# because they didn't support C# 4.0 or higher for a long time.

**Comment by Warlaan** [comment]
Looking at the current implementation it seems like a reasonable solution to store the previous GDScriptFunctionState in every returned GDScriptFunctionState. So when a GDScriptFunctionState determines that it should not signal its "completed" signal in [gdscript_function.cpp:1539](https://github.com/godotengine/godot/blob/d345712547f39df1375207d3543c00ef51c8e6bf/modules/gdscript/gdscript_function.cpp#L1539) it should store a pointer to itself in the new GDScriptFunctionState it just deteced. That way when the coroutine ends and completed is true the "completed" signal can be emitted not only for the current state but also for all previous ones until the first one is reached where the previousState-pointer is empty.

**Comment by Warlaan** [comment]
I submitted a PR with the solution I described in the last response. 

Feedback welcome, especially since it triggers a warning that I don't quite understand (details in the PR description).

**Comment by vnen** [comment]
Wasn't this solved by the PR?
